logdna/resource_stream_config.go

```diff
@@ -0,0 +1,175 @@
+package logdna
+
+import (
+	"context"
+	"encoding/json"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+const streamConfigID = "stream"
+
+type streamConfig struct {
+	Status   string   `json:"status,omitempty"`
+	Brokers  []string `json:"brokers"`
+	Topic    string   `json:"topic"`
+	User     string   `json:"user"`
+	Password string   `json:"password"`
+}
+
+func resourceStreamConfigCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	pc := m.(*providerConfig)
+
+	config := streamConfig{
+		Brokers:  listToStrings(d.Get("brokers").([]interface{})),
+		Topic:    d.Get("topic").(string),
+		User:     d.Get("user").(string),
+		Password: d.Get("password").(string),
+	}
+
+	req := newRequestConfig(
+		pc,
+		"POST",
+		"/v1/config/stream",
+		config,
+	)
+
+	body, err := req.MakeRequest()
+
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	created := streamConfig{}
+	err = json.Unmarshal(body, &created)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(streamConfigID)
+
+	return resourceStreamConfigRead(ctx, d, m)
```

Comment:
Just for my understanding, is the workflow in the function here something along the lines of:

1. Pull the provider config into `pc` (assuming this is to grab the service key for the POST request to our Config API)
2. Assign `config` to the intended event stream setup defined as a resource in the local `.tf` configuration file
3. Connect LogDNA to the relevant brokers by sending the config defined above in a POST request to our Config API
4. Check whether the stream setup was successfully created by unmarshaling the response JSON into `created`
5. Set local stream config's resource ID to the unary "uid" so we reconcile its state in the returned GET (read) method

I'm not entirely sure if steps `#2`, `#4`, and `#5` are accurately capturing what's happening. I'm also a bit confused with step `#5`, what's the purpose of reading the remote resource that we presumably just created in this method?

Reply:
overall that's pretty accurate, just a few things:

> 1. Pull the provider config into `pc` (assuming this is to grab the service key for the POST request to our Config API)

yes, this contains general configuration and a pre-wired http client (so there's only a single client in play, and common configuration for the client is defined only once) https://github.com/logdna/terraform-provider-logdna/blob/854b57bcffb9f0e8adda8a9418e86dcae08c2a80/logdna/provider.go#L43-L45

> 2. Assign `config` to the intended event stream setup defined as a resource in the local `.tf` configuration file

this takes the config supplied in the `tf` file, typechecks it, and assigns it to a new struct

> 3. Connect LogDNA to the relevant brokers by sending the config defined above in a POST request to our Config API
> 4. Check whether the stream setup was successfully created by unmarshaling the response JSON into `created`

yeah the `created` struct isn't used, but as a sanity check we unmarshal the response (which is just the JSON representation of the config that was saved).

if the response code was not 200, we would get an error above. this block verifies that the response _shape_ is what we expect.

> 5. Set local stream config's resource ID to the unary "uid" so we reconcile its state in the returned GET (read) method
> 
> I'm not entirely sure if steps `#2`, `#4`, and `#5` are accurately capturing what's happening. I'm also a bit confused with step `#5`, what's the purpose of reading the remote resource that we presumably just created in this method?

The last step here of `return resourceStreamConfigRead(...)` is more of a general best-practice with TF resources, as there are sometimes computed properties that you'd need to refetch from the API. These new endpoints do make it a bit easier since the created resource is returned directly.

In the case of the stream config, there's no resource ID - there can only be one "config", so a static ID is used instead. You're right that we don't necessarily _need_ to re-read for this.

Reply:
I updated the `create` functions for both new resources to skip the extra `read`, since the full resource is returned in full from a `POST`. Technically the same could be done for the `update` but I think the pattern of update-then-read is a a good one to keep there.